### PR TITLE
Cody Web record chat id with event logs

### LINKED ISSUE #55492: Cody Web record chat id with event logs.
For all the cody events related to a chat, record the chat id, so that we can analyze user activity per chat. 

Events: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/util/constants.ts

### PR DESCRIPTION
**Pull request by st0nebraker**
Closes #55492, adds chat id for every Cody Web chat event log.

## Test plan
- CI/CD

### CODE REVIEW

**Review by thenamankumar — APPROVED** [review]

**Review by courier-new — COMMENTED** [review]
Yay! Thanks Becca! Three non-blocking thoughts:

- Ought we to just add a basic utility function that wraps `eventLogger.log()` to use in these places? Like
	```ts
	const logTranscriptEvent = (eventLabel: string, chatId: string) => eventLogger.log(eventLabel, { chatId })
	```
	That would help encourage inclusion of the chat id in future events that we log, too, and prevent any accidental divergence in the property name.
- `chatId` or `transcriptId`? I know, I know, does it really matter? But I couldn't help but notice the value is always `transcript.id`, so why not align on that naming to reduce terminology overloading?
- It kinda sucks that `transcript` is almost always nullable wherever we're logging its ID. This is obviously a problem outside of the context of event logging, too, but I'm just curious what all our options are here. Do you have much of an understanding about when `transcript` would be `null` in these cases? Does it even make sense to run these functions or log these events in the first place if the `transcript` is `null`? No action required here, just want to understand the situation here a bit better. I'd hate to end up with a bunch of "bad" event data that's missing transcript IDs because we didn't consider a legitimate case of a `null` transcript in one of these places. 😅

**Comment by st0nebraker** [comment]
Apologies! I had this PR on auto-merge, so it merged before I could get to your comments. Happy to make a follow up PR for this feedback 😄 

> * It kinda sucks that `transcript` is almost always nullable wherever we're logging its ID. This is obviously a problem outside of the context of event logging, too, but I'm just curious what all our options are here. Do you have much of an understanding about when `transcript` would be `null` in these cases? Does it even make sense to run these functions or log these events in the first place if the `transcript` is `null`? No action required here, just want to understand the situation here a bit better. I'd hate to end up with a bunch of "bad" event data that's missing transcript IDs because we didn't consider a legitimate case of a `null` transcript in one of these places. 😅

On it's definition it's nullable ([here](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/cody/-/blob/lib/shared/src/chat/useClient.ts?L94)), however, I've noticed it always has a value after it instantiates. You'll notice a brand new Cody interaction always has a `transcript` value. So I don't believe we'll be recording any "bad" data, but we should refactor this and make `transcript` non-nullable if possible. @thenamankumar do you have any added context here why we wouldn't be able to do that?

> * Ought we to just add a basic utility function that wraps `eventLogger.log()` to use in these places? Like
>   ```ts
>    const logTranscriptEvent = (eventLabel: string, chatId: string) => eventLogger.log(eventLabel, { chatId })
>   ```
>   
>   That would help encourage inclusion of the chat id in future events that we log, too, and prevent any accidental divergence in the property name.

That's a good idea! I'm worried it would get lost and not used in the future though. What do you think? How/where should I add the utility function so it's clear enough & obvious to use for future devs?

> * `chatId` or `transcriptId`? I know, I know, does it really matter? But I couldn't help but notice the value is always `transcript.id`, so why not align on that naming to reduce terminology overloading?

Good point, addressed in a follow-up #55937

File: client/web/src/cody/useCodyChat.tsx [hunk 1]
```diff
@@ -188,7 +188,8 @@ export const useCodyChat = ({
             return
         }
 
-        eventLogger.log(EventName.CODY_CHAT_HISTORY_CLEARED)
+        const chatIds = transcriptHistory.map(chat => chat.id)
+        eventLogger.log(EventName.CODY_CHAT_HISTORY_CLEARED, { chatIds })
```
[the remark this answers is not in the document]
Comment (st0nebraker) [reply client/web/src/cody/useCodyChat.tsx:192]: Fair point, addressed in #55937!

**Comment by courier-new** [comment]
> That's a good idea! I'm worried it would get lost and not used in the future though. What do you think? How/where should I add the utility function so it's clear enough & obvious to use for future devs?

Yeah that's a great question. My gut feeling is, since you were having to pass `transcript` around everywhere to log this event anyway, what if we define it in the same place `transcript` comes from, aka the `CodyChatStore`? That actually has another added benefit, which is that we'll already have `chatId` available, too, and won't even _need_ to pass `transcript` around anymore!

```ts
// Add the utility function to the return type
export interface CodyChatStore {
	...
	logTranscriptEvent: (eventLabel: string) => void
}

export const useCodyChat = ({ ... }): CodyChatStore => {
	...
	// Within here, `transcript` itself is already defined and managed
	...
	const logTranscriptEvent = (eventLabel: string) => {
		// `transcript` is now already available. At least now, even if it's null,
		// we only have to deal with it being null in a single place. 🤷‍♀️
		if (!transcript) {
			return
		}
		eventLogger.log(eventLabel, { chatId: transcript.id })
	}
	...
	return {
		...
		logTranscriptEvent
	}
}
```

I think that should make it fairly discoverable and obvious how to use because it'll be available anywhere else we'd normally be handling Cody chat interactions. What do you think? 🙂

**Comment by st0nebraker** [comment]
> > That's a good idea! I'm worried it would get lost and not used in the future though. What do you think? How/where should I add the utility function so it's clear enough & obvious to use for future devs?
> 
> Yeah that's a great question. My gut feeling is, since you were having to pass `transcript` around everywhere to log this event anyway, what if we define it in the same place `transcript` comes from, aka the `CodyChatStore`? That actually has another added benefit, which is that we'll already have `chatId` available, too, and won't even _need_ to pass `transcript` around anymore!
> 
> ```ts
> // Add the utility function to the return type
> export interface CodyChatStore {
> 	...
> 	logTranscriptEvent: (eventLabel: string) => void
> }
> 
> export const useCodyChat = ({ ... }): CodyChatStore => {
> 	...
> 	// Within here, `transcript` itself is already defined and managed
> 	...
> 	const logTranscriptEvent = (eventLabel: string) => {
> 		// `transcript` is now already available. At least now, even if it's null,
> 		// we only have to deal with it being null in a single place. 🤷‍♀️
> 		if (!transcript) {
> 			return
> 		}
> 		eventLogger.log(eventLabel, { chatId: transcript.id })
> 	}
> 	...
> 	return {
> 		...
> 		logTranscriptEvent
> 	}
> }
> ```
> 
> I think that should make it fairly discoverable and obvious how to use because it'll be available anywhere else we'd normally be handling Cody chat interactions. What do you think? 🙂

Brilliant!! Thank you Kelli. This makes total sense to me. I have that follow up PR ready [here](https://github.com/sourcegraph/sourcegraph/pull/55937) :)

**Pull request merged**